### DISCUSSION TITLE
Nim style name equality?

### DISCUSSION THREAD

**Discussion by joakim** [joakim/kesh#3] [Ideas]
## Suggestion

Extend [na's naming rules](https://github.com/kesh-lang/na#names) to also support `camelCase` and `PascalCase`.

Unlike [Nim's identifier equality](https://nim-lang.org/docs/manual.html#lexical-analysis-identifier-equality), have all names be case insensitive, no special exception for the first character (kesh doesn't have classes).

In other words, the following being equal:

```lua
hello-world
hello_world
HELLO-WORLD
HELLO_WORLD
helloWorld
HelloWorld
```

### Issues

There are some issues with this approach:

When transpiled to ECMAScript, names should be in ECMAScript's idiomatic [camelCase](https://en.wikipedia.org/wiki/Letter_case#Camel_case), at least by default. That means that, unlike Nim, `helloworld` (no word separators) can't be evaluated as identical to the variants listed above, as the transpiler must be able to separate the words.

More troubling, any uppercase abbreviations within names would lose their case. E.g. `encode-uri` would become `encodeUri` instead of `encodeURI`. The same goes for ECMAScript classes and constructors, starting with an uppercase letter, like `Error`, `Date`, `Math`, etc. The compiler would need to keep a lookup table of normalized names of ECMAScript globals and their fields, e.g. `encode-uri-component` should transpile to `encodeURIComponent()`, not the incorrect `encodeUriComponent()`.

External JavaScript/TypeScript code would also have this issue. Any `snake_cased` names spotted in the wild too. When importing incompatible names from a non-kesh module, one _could_ rename it using a string to reference exported values as named by the module.

```lua
import ['RouterLink' as router-link]: 'vue-router'
```

For direct field access on objects, one could similarly use a string.

```lua
import vue-router: 'vue-router'

router-link: vue-router.'RouterLink'
```

Is it worth the trouble?

## Alternative suggestion

An alternative and much simpler approach is to transpile only kesh idiomatic kebab-cased names to ES idiomatic camelCase. All other forms are transpiled verbatim.

```lua
kebab-case  --> kebabCase
Kebab-case  --> KebabCase
camelCase   --> camelCase
PascalCase  --> PascalCase
snake_case  --> snake_case
SCREAMING_SNAKE_CASE  --> SCREAMING_SNAKE_CASE
sPoNgEbOb_sNaKe_cAsE  --> sPoNgEbOb_sNaKe_cAsE
```

Basically, `-` becomes a modifier uppercasing the following character.

```lua
encode-URI    --> encodeURI
encode-URI-component  --> encodeURIComponent
URI-error     --> URIError
JSON-parser   --> JSONParser
JSON          --> JSON
is-NaN        --> isNaN
NaN           --> NaN
```

### Issues

None?

## Conclusion

Simple is better. The alternative suggestion it is.
